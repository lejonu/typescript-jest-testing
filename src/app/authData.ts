export interface IAuthData {
  usernameToLower: string;
  usernameCharacteres: Array<string>;
  userDetails: Object | undefined;
  isAuthenticated: boolean;
}
