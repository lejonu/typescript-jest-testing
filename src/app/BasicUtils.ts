import { IAuthData } from "./authData";

export function product(a: number, b: number): number {
  return a * b;
}

export function authenticateUser(
  username: string,
  password: string
): IAuthData {
  const authStatus = username === "deveLOPER" && password === "dev";
  return {
    usernameToLower: username.toLowerCase(),
    usernameCharacteres: username.split(""),
    userDetails: {},
    isAuthenticated: authStatus,
  };
}

export class UserNameToLowercase {
  public toLower(username: string): string {
    if (username.trim() === "") {
      throw new Error("Invalid username!");
    }
    return username.toLowerCase();
  }
}
