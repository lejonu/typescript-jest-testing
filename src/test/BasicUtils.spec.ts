import { authenticateUser, product } from "../app/BasicUtils";

describe("BasicUtils test suite", () => {
  it("returns the product of 3 an 2", () => {
    // arrange
    const sut = product;

    // acting
    const actual = sut(3, 2);

    // asserting
    expect(actual).toBe(6);
    expect(actual).toEqual(6);
    expect(actual).toBeLessThan(10);
    expect(actual).toBeGreaterThan(4);
  });

  // Nested describe
  describe("Authentication test suite", () => {
    it("return the lower case of a valid user", () => {
      const sut = authenticateUser;
      const actual = sut("deveLOPER", "dev");
      expect(actual.usernameToLower).toBe("developer");
    });

    it("return the username characters of a valid user", () => {
      const sut = authenticateUser;
      const actual = sut("deveLOPER", "dev");
      // testing arrays or objects use toEqual
      expect(actual.usernameCharacteres).toEqual([
        "d",
        "e",
        "v",
        "e",
        "L",
        "O",
        "P",
        "E",
        "R",
      ]);

      // Check if the array contains a particular character
      expect(actual.usernameCharacteres).toContain("R");
    });

    it("return the username shuffled characters of a valid user", () => {
      const sut = authenticateUser;
      const actual = sut("deveLOPER", "dev");
      // testing arrays or objects use toEqual
      // Check if the array contains a particular character
      expect(actual.usernameCharacteres).toEqual(
        expect.arrayContaining(["L", "O", "P", "E", "R", "d", "e", "v", "e"])
      );
    });

    it("return the username details object from a valid user", () => {
      const sut = authenticateUser;
      const actual = sut("deveLOPER", "dev");
      // testing arrays or objects use toEqual
      expect(actual.userDetails).toEqual({});
    });

    it("return if username details object is defined", () => {
      const sut = authenticateUser;
      const actual = sut("deveLOPER", "dev");

      expect(actual.userDetails).toBeDefined();
      expect(actual.userDetails).not.toBeUndefined();
      expect(actual.userDetails).toBeTruthy();
      expect(actual.userDetails).not.toBeFalsy();
    });

    it("return if user is authenticated", () => {
      const sut = authenticateUser;
      const actual = sut("deveLOPER", "dev");

      expect(actual.isAuthenticated).toBeTruthy();
    });
  });

  //   //only runs just a block of test
  //   it.only("User authentication", () => {
  //     const sut = authenticateUser;
  //     const actual = sut("deveLOPER", "dev");
  //     expect(actual.usernameToLower).toBe("developer");
  //   });
});
