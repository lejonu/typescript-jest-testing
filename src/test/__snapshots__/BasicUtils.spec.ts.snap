// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicUtils test suite class UserNameToLowercase test suite throws an error for an invalid username: Invalid username! 1`] = `"Invalid username!"`;
